### PR TITLE
CRM-21289 Implementation

### PR DESCRIPTION
**Pull request by bjendres**
Overview
----------------------------------------
This PR implements the new hook outlined in [CRM-21289](https://issues.civicrm.org/jira/browse/CRM-21289), along with some unit tests. We'll create a PR to the documentation as soon as this is accepted.

Before
----------------------------------------
Hook doesn't exist.

After
----------------------------------------
Hook does exist :)

Technical Details
----------------------------------------
The hook can be used to create any custom generated ID (i.e. non DB related). We also plan to use it with extensions (e.g. mandate references in CiviSEPA). Currently it has four applications:
 * ``invoice_number`` in CiviContribute
 * ``creditnote_id`` in CiviContribute
 * ``external_identfier`` for contacts
 * ``external_identfier`` for campaigns

Comments
----------------------------------------
One of the main triggers for the creation of this hook was the incomplete implementation of invoice ids (always equals the contribution id), which violates several requirements for invoicing - at least in Germany and Belgium.

### CODE REVIEW

File: tests/phpunit/CRM/Utils/Hook/GenerateIdentifierTest.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
+<?php
+/**
+ * Class CRM_Utils_Hook_Generate_Identifier_Test
+ *
+ * Tests the civicrm_generateIdentifier hook (see CRM-21289)
+ *
+ * @group headless
+ */
+class CRM_Utils_Hook_Generate_Identifier_Test extends CiviUnitTestCase {
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Utils/Hook/GenerateIdentifierTest.php]: ohh look at all these pretty unit tests :-)

Comment (systopia) [reply tests/phpunit/CRM/Utils/Hook/GenerateIdentifierTest.php]: :-)

**Comment by systopia** [comment]
Corresponding docs PR is here: https://github.com/civicrm/civicrm-dev-docs/pull/423

**Comment by KarinG** [comment]
I really like adding more hooks to give people more options to use CiviCRM to make it work (better) for them. Nice tests - I'm going to steal them as an example to test the additional_non_deductible_amount hook in our CDN Tax Receipts Extension... The Documentation docs PR contains example code - are you guys trying to score bonus points? :-)

**Comment by colemanw** [comment]
I'm confused about the specificity of this hook. It is essentially a presave hook, which we already have. In fact we have a couple of them - one at the BAO level and one at the DAO level. What about this hook makes it specific to external identifiers? Why are the generic presave hooks not adequate?

**Comment by systopia** [comment]
> I'm confused about the specificity of this hook. It is essentially a presave hook, which we already have. In fact we have a couple of them - one at the BAO level and one at the DAO level. What about this hook makes it specific to external identifiers? Why are the generic presave hooks not adequate?

@colemanw: That's a point that @lcdservices raised as well. But while the two ``external_identifier`` could be implemented using the pre-hooks, the ``invoice_number`` and the ``creditnote_id`` certainly cannot - that is: not without a mayor refactoring of the invoicing code.

We just added in the two ``external_identifier`` because we thought the
idea of the hook should be to "Influence the generation of (unique) IDs
that are communicated to the 'outside world'". In my eyes that is a very useful hook, that I would be using in my CiviSEPA extension as well, where similar unique references are generated and sent to third parties.

**Comment by lcdservices** [comment]
@systopia I was surprised that you said it was not possible to impact those two fields via the pre hook, so I put together a quick extension to test it. I don't know the specifics of your implementation, but it certainly is possible to construct values with that hook:

```
function test_civicrm_pre($op, $objectName, $id, &$params) {
  if ($op == 'create' && $objectName == 'Contribution') {
    $rnd = rand();
    $params['invoice_number'] = "Invoice-{$rnd}";
  }

  if ($op == 'edit' &&
    $objectName == 'Contribution' &&
    !empty($params['creditnote_id']) &&
    strpos($params['creditnote_id'], 'CN_') !== FALSE
  ) {
    $rnd = rand();
    $params['creditnote_id'] = "Credit-{$rnd}";
  }
}
```

I store the invoice_number on contribution creation, as I suspect that's what you'd want. You'd implement a function to get the next ID in your series where I just used a random number.

The credit note ID might be a little trickier -- that value is generated when you refund the contribution, so I conditioned on the existence of a value where the value is in the standard Civi format (prefix = CN_) and then set it to my custom value.

I just want to make sure we explore the feasibility of using existing hooks if possible, especially since the proposed hook seems so narrow in scope.

**Comment by systopia** [comment]
@lcdservices: The problem with invoicing is that the ``invoice_number`` cannot be created along with the contribution, but only when the *invoice* is generated - otherwise you'll never be able to create sequentially numbered invoices (in the order of the issue date).

The same holds for the ``creditnote_id``: only when a credit note is generated, an id can be created alongside - everything else breaks sequential numbering. This would also be the case for the usage of this hook in my extensions.

Have a look at where the new hooks are placed with this PR. Let me know if this doesn't answer your question.

**Comment by lcdservices** [comment]
@systopia So I think we're looking at things a bit differently. I would argue that contribution records *are* effectively invoices. A completed contribution simply means the invoice is paid and closed; a pending contribution means it is open. You are considering the generation of a PDF invoice as being the creation of the invoice record. In reality, all that is really doing is running the contribution record through the message template system to create the PDF file -- it doesn't actually create something called an "invoice" record.

Anyway... because that PDF invoice is just a msg template, it passes through the alterMailContent and alterMailParams hooks. In particular, alterMailParams passes all the variables that can be used in the template markup. So if that is your trigger (PDF generation) for indicating an invoice is created, you could set the invoice_number there. I'm less familiar with the credit note process -- I know from a data standpoint, the credit ID is set when a contrib is marked refunded (which was my comment above).

**Comment by systopia** [comment]
> @lcdservices: So I think we're looking at things a bit differently. I would argue that contribution records are effectively invoices. A completed contribution simply means the invoice is paid and closed; [...]

That's fair enough, but our client's interpretation is a different one. Their national regulations require a gap-less, sequential numbering in the order of issuing. They also don't need invoices for all financial types, just some. You can see how they wouldn't agree with your interpretation. But that's the point of the hook: everyone can implement it the way they need.

With respect to the ``alterMailParams`` hook: I'm fairly sure that doesn't happen in this case. You can believe me that I've spent quite some time trying to solve this without changes to core, because I know how much time this process can take.

On top of that, from a usability point of view, I would greatly prefer a clear ``generateIdentifier`` hook over telling people they would have to use ``alterMailParams`` to generate invoice numbers (if that turned out to be viable after all).

**Comment by lcdservices** [comment]
`alterMailParams` and `alterMailContent` are definitely hit when the invoice is created. I use a combination of them on several sites to completely rewrite the default invoice template (which has horrible markup) for several sites (I find it easier to do that via code than through the message template UI). I agree it _feels_ a little odd to use those hooks, because of how they're named. But I think they are a feasible solution. I can provide sample code if that's helpful.

I appreciate the desire to target those values more intentionally. I don't have any issues with the PR code, per se -- just the reservations voiced above (customizations could be handled in other ways; implementation alters data in some contexts and not in others, etc). I think @colemanw needs to decide if that intention warrants the hook even if there are other solutions for targeting those fields.

**Comment by monishdeb** [comment]
> @lcdservices: The problem with invoicing is that the invoice_number cannot be created along with the contribution, but only when the invoice is generated - otherwise you'll never be able to create sequentially numbered invoices (in the order of the issue date). 

Agree, but there was a PR  https://github.com/civicrm/civicrm-core/pull/10681 of Pradeep which allows us to generate and store invoice_number when contribution created/updated.

**Comment by lcdservices** [comment]
@monishdeb where @systopia said "cannot be created" what he was really saying was that his client wants the invoice_number created during PDF generation -- not during contribution creation. He wasn't really saying that it was not possible, but that his desired approach (via customization) is to create it at a different point in the process.

**Comment by monishdeb** [comment]
@lcdservices sorry if I meant it otherwise. I just mentioned the PR to notify the changes Pradeep did, so it might be of any help here. But later realised it's not relevant here, my bad :(

**Comment by systopia** [comment]
So where are we now? 

I would maintain that even if we could somehow manage to generate ``invoice_number`` and ``creditnote_id`` within the ``alterMailParams`` / ``alterMailContent`` hooks, the solution presented in this PR is a lot clearer and way more intuitive. I would argue that it is a good improvement to the CiviCRM core, has no side-effects (as it's a new hook), and I would go ahead and ask @totten to merge it.

@lcdservices, @colemanw: I think @KarinG has given her "thumbs up", what's your verdict?

**Comment by colemanw** [comment]
I still think this hook is not necessary. The use-case you describe is a nonstandard one, but it sounds like existing hooks can be utilized to make the system behave the way you want.

**Comment by systopia** [comment]
@eileenmcnaughton Well, it's not really off my table, I just can't find the time right now to convince you guys. I'll be back :) Eventually.

**Comment by eileenmcnaughton** [comment]
@systopia - fair enough - I think it's good to close these while they are not moving & re-open them when you have time to petition for it (or lure CT to a sprint in UK and then corner them :-)

**Pull request closed without merging**